### PR TITLE
[go/sdk-gen] Generating SDK with generics

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
# Description

This PR implements generating the generic variant of a go SDK from Pulumi schemas. Currently the idea is to generate a directory `x` inside the root directory of the go SDK which will contain the same SDK except using generics and generating far less code than its current counter part. 

Also implements an enum option `$.language.go.generics` which can be set to the following:
 - `none` is the default which maintains the current behavior that generates legacy types without generics
 -  `side-by-side` generates the generics sdk variant alongside the current sdk under directory `x`
 -  `generics-only` generates only the new sdk with generics at the root of the package
 
Still a bunch of things to do:
 - [x] Generating `InvokeResult]Output` type from `Output[InvokeResult]` and generating accessor methods for it
 - [x] Generating default values for types and using the `pulumix` subpackage to do so
 - [x] Generating generic SDK variants for all test schemas we have and making sure they compile (currently only testing `simple-resource-schema` as shown below)
 - [x] Account for plain inputs for components
 - [x] Combine pulumix.Join with pulumix.Apply to generate resource accessor methods
     - [x] Problem with `GPtrOutput[T]`  and `ArrayOutput[T]` being unwrapped to `Output[*T]` and `Output[[]T]`
 - [x] Remove excess untyped container types from generated enums
 - [x] Fix default values for resource methods with lifted single return value
 - [x] Secret properties

Currently the following test schemas have opted for `generics: "side-by-side"`:
 - [x] output-funcs
 - [x] simple-enum-schema
 - [x] secrets
 - [x] simple-plain-schema
 - [x] plain-and-default

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/provider.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	return reflect.TypeOf((*Provider)(nil)).Elem()
+}
+
+func (o ProviderOutput) ToProviderOutput() ProviderOutput {
+	return o
+}
+
+func (o ProviderOutput) ToProviderOutputWithContext(ctx context.Context) ProviderOutput {
+	return o
+}
```
Comment (justinvp) [inline pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/provider.go:58]: For the generic variant of these output types, we likely don't need these two methods. Is there any reason to keep them?

File: pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/pulumiTypes.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+func (o ContainerOutput) ToContainerOutputWithContext(ctx context.Context) ContainerOutput {
+	return o
+}
+
+func (o ContainerOutput) ToOutput(ctx context.Context) pulumix.Output[Container] {
+	return pulumix.Output[Container]{
+		OutputState: o.OutputState,
+	}
+}
```
Comment (justinvp) [inline pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/pulumiTypes.go:89]: It seems like something fishy is going on here. Let's not block on it for this PR, but would like to figure it out subsequently.

1. For the generic variants, I'm not sure there's a need for `ToFooOutput` and `ToFooOutputWithContext` methods.
2. As I mentioned elsewhere, I think we can use a type definition for the Output type, which I think avoids the need to generate `ElementType` and `ToOutput` methods.
3. `(i *ContainerArgs) ToOutput(ctx context.Context) pulumix.Output[*ContainerArgs]` looks suspect to me. I understand why it's that way at the moment to get things to work, but I think we need to revisit this. It seems like the signature should be something like:

    ```go
    (i *ContainerArgs) ToOutput(ctx context.Context) pulumix.Output[Container]
    ```

    or:

    ```go
    (i ContainerArgs) ToOutput(ctx context.Context) pulumix.Output[Container]
    ```


```suggestion
func (i *ContainerArgs) ToOutput(ctx context.Context) pulumix.Output[*ContainerArgs] {
	return pulumix.Val(i)
}

type ContainerOutput pulumix.Output[Container]
```

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/provider.go (hunk 1 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/provider.go:58]: To be honest I am not entirely sure if they are needed. I know the marshalling logic in the core SDK sometimes needs to call `To{TypeName}OutputWithContext(...)` using reflection but whether it is still the case with the generic types, is unclear

File: pkg/codegen/testing/test/testdata/output-funcs/go/mypkg/x/funcWithAllOptionalInputs.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+func (FuncWithAllOptionalInputsResultOutput) ElementType() reflect.Type {
+	return reflect.TypeOf((*FuncWithAllOptionalInputsResult)(nil)).Elem()
+}
+
+func (o FuncWithAllOptionalInputsResultOutput) ToOutput(context.Context) pulumix.Output[*FuncWithAllOptionalInputsResult] {
+	return pulumix.Output[*FuncWithAllOptionalInputsResult]{
+		OutputState: o.OutputState,
+	}
+}
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/testdata/output-funcs/go/mypkg/x/funcWithAllOptionalInputs.go:74]: Let's try do that in a follow up PR better than having this one get even bigger 😅

File: pkg/codegen/go/gen.go [hunk 4]
```diff
@@ -2097,63 +2486,120 @@ func (pkg *pkgContext) genResource(w io.Writer, r *schema.Resource, generateReso
...
-			name, outputType, pkg.fieldName(r, p), outputType)
+		if !useGenericVariant {
+			fmt.Fprintf(w, "\treturn o.ApplyT(func (v *%s) %s { return v.%s }).(%s)\n",
+				name, outputType, pkg.fieldName(r, p), outputType)
+		} else {
+			needsCast := strings.HasPrefix(outputType, "pulumix.ArrayOutput") ||
+				strings.HasPrefix(outputType, "pulumix.MapOutput") ||
+				strings.HasPrefix(outputType, "pulumix.GPtrOutput") ||
+				strings.HasPrefix(outputType, "pulumix.GArrayOutput") ||
+				strings.HasPrefix(outputType, "pulumix.GMapOutput")
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/go/gen.go:2561]: Done

File: pkg/codegen/go/gen.go [hunk 5]
```diff
@@ -2322,11 +2794,123 @@ func (pkg *pkgContext) functionResultTypeName(f *schema.Function) string {
...
+	originalResultTypeName := pkg.functionResultTypeName(f)
+	resultTypeName := fmt.Sprintf("%sOutput", originalResultTypeName)
+
+	code := `
+func ${fn}Output(ctx *pulumi.Context, args ${fn}OutputArgs, opts ...pulumi.InvokeOption) ${outputType} {
+	outputResult := pulumix.ApplyErr[*${fn}Args](args.ToOutput(), func(plainArgs *${fn}Args) (*${fn}Result, error) {
+		return ${fn}(ctx, plainArgs, opts...)
+	})
+
+    return pulumix.Cast[${outputType}, *${fn}Result](outputResult)
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/go/gen.go:2809]: Done

File: pkg/codegen/go/gen.go [hunk 6]
```diff
@@ -824,6 +913,143 @@ func (pkg *pkgContext) outputTypeImpl(t schema.Type) string {
...
+			return fmt.Sprintf("map[string]%s", elementType), false
+		case *schema.UnionType:
+			for _, e := range schemaType.ElementTypes {
+				if enumType, ok := e.(*schema.EnumType); ok {
+					return pkg.genericElementType(enumType.ElementType)
+				}
+			}
+			return "interface{}", true
+		default:
+			return "interface{}", true
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/go/gen.go:963]: Done

File: pkg/codegen/go/gen.go [hunk 7]
```diff
@@ -824,6 +913,143 @@ func (pkg *pkgContext) outputTypeImpl(t schema.Type) string {
...
+		}
+	}
+}
+
+// genericOutputTypeImpl is similar to outputTypeImpl, but it generates the output type variants.
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/go/gen.go:968]: Nice catch! adjusted

File: pkg/codegen/go/gen_crd2pulumi.go [hunk 8]
```diff
@@ -40,18 +40,18 @@ func CRDTypes(tool string, pkg *schema.Package) (map[string]*bytes.Buffer, error
...
 
 		if len(pkg.types) > 0 {
 			for _, t := range pkg.types {
-				if err := pkg.genType(buffer, t); err != nil {
+				if err := pkg.genType(buffer, t, false); err != nil {
 					return nil, err
 				}
 			}
-			pkg.genTypeRegistrations(buffer, pkg.types)
+			pkg.genTypeRegistrations(buffer, pkg.types, false /* usingGenericTypes */)
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/go/gen_crd2pulumi.go:54]: It is the actual name, just checked it again to be sure

File: pkg/codegen/go/importer.go [hunk 9]
```diff
@@ -91,6 +91,13 @@ type GoPackageInfo struct {
 	// InternalDependencies are blank imports that are emitted in the SDK so that `go mod tidy` does not remove the
 	// associated module dependencies from the SDK's go.mod.
 	InternalDependencies []string `json:"internalDependencies,omitempty"`
+
+	// Specifies how to handle generating a variant of the SDK that uses generics.
+	// Allowed values are the following:
+	// - "none" (default): do not generate a generics variant of the SDK
+	// - "side-by-side": generate a side-by-side generics variant of the SDK under the x subdirectory
+	// - "only-generics": generate a generics variant of the SDK only
+	GenericsVariantSetting string `json:"genericsVariantSetting,omitempty"`
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/go/importer.go:100]: Done

File: pkg/codegen/testing/test/testdata/go-plain-ref-repro/go/repro/ecs/fargateTaskDefinition.go [hunk 10]
```diff
@@ -41,8 +41,8 @@ type fargateTaskDefinitionArgs struct {
 
 // The set of arguments for constructing a FargateTaskDefinition resource.
 type FargateTaskDefinitionArgs struct {
-	Container  *TaskDefinitionContainerDefinitionArgs
-	Containers map[string]TaskDefinitionContainerDefinitionArgs
+	Container  *TaskDefinitionContainerDefinition
+	Containers map[string]TaskDefinitionContainerDefinition
```
[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/testdata/go-plain-ref-repro/go/repro/ecs/fargateTaskDefinition.go:45]: This is the only thing left

[the remark this answers is not in the document]
Comment (Zaid-Ajaj) [reply pkg/codegen/testing/test/testdata/go-plain-ref-repro/go/repro/ecs/fargateTaskDefinition.go]: Fixed ✅ it was a small thing but it was buried under a lot of logic that simplifies types

File: pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/provider.go (hunk 1 above)
Comment (justinvp) [reply pkg/codegen/testing/test/testdata/simple-enum-schema/go/plant/x/provider.go:58]: Good point. Let's follow-up on this separately.

File: pkg/codegen/go/gen_crd2pulumi.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (justinvp) [reply pkg/codegen/go/gen_crd2pulumi.go:54]: Ah, sorry, I must have been looking at another function.

**Pull request merged**